Libraries/MiscDrivers/CODEC/max9867.c

```diff
@@ -0,0 +1,431 @@
+}
+
+int max9867_linein_gain(int left, int right)
+{
+  int err;
+  uint8_t ligl, ligr;
+
+  ligl = (24 - left) / 2;
+  ligr = (24 - right) / 2;
+
+  if ((err = reg_update(MAX9867_0E_LVL_LINE_IN_LEFT, 0x0F, ligl)))
+    return err;
+
+  return reg_update(MAX9867_0F_LVL_LINE_IN_RIGHT, 0x0F, ligr);
+}
+
+int max9867_linein_vol_fix(int fix)
+{
+  return reg_update(MAX9867_09_DAI_CLOCK, 0x01, !!fix << 4);
+}
+
+int max9867_playback_mute(int left, int right)
+{
+  int err;
+
+  if ((err = reg_update(MAX9867_10_VOL_LEFT, 0x40, !!left << 6)))
+    return err;
+
+  return reg_update(MAX9867_11_VOL_RIGHT, 0x40, !!right << 6);
+}
+
+static uint8_t map_vol(float vol)
+{
+  int tmp;
+  uint8_t res;
+
+  tmp = vol * 10;
+
+  if (tmp >= 30)
+    res = (60 - tmp) / 5;
+  else if (tmp >= -60)
+    res = (2 - (int)vol) + 7;
+  else if (tmp >= -220)
+    res = (8 - (int)vol) / 2 + 8;
+  else if (tmp >= -860)
+    res = (26 - (int)vol) / 4 + 11;
+  else
+    res = 0x3F;
+
+  return res;
+}
+
+int max9867_playback_volume(float left, float right)
+{
+  int err;
+  uint8_t voll, volr;
+
+  voll = map_vol(left);
+  volr = map_vol(right);
+
+  if ((err = reg_update(MAX9867_10_VOL_LEFT, 0x3F, voll)))
+    return err;
+
+  return reg_update(MAX9867_11_VOL_RIGHT, 0x3F, volr);
+}
+
+int max9867_microphone_preamp(mic_pa_gain_t left, mic_pa_gain_t right)
+{
+  int err;
+
+  if ((err = reg_update(MAX9867_12_MIC_GAIN_LEFT, 0x60, left << 5)))
+    return err;
+
+  return reg_update(MAX9867_13_MIC_GAIN_RIGHT, 0x60, right << 5);
+}
+
+int max9867_microphone_gain(int left, int right)
+{
+  int err;
+  uint8_t pgaml, pgamr;
+
+  pgaml = 20 - left;
+  pgamr = 20 - right;
+
+  if ((err = reg_update(MAX9867_12_MIC_GAIN_LEFT, 0x1F, pgaml)))
+    return err;
+
+  return reg_update(MAX9867_13_MIC_GAIN_RIGHT, 0x1F, pgamr);
+}
+
+int max9867_adc_input(adc_input_t left, adc_input_t right)
+{
+  return reg_update(MAX9867_14_ADC_INPUT, 0xF0, (left << 6) | (right << 4));
+}
+
+int max9867_microphone_clock(mic_clk_t mclk)
+{
+  return reg_update(MAX9867_15_MIC, 0xC0, mclk << 6);
+}
+
+int max9867_microphone_enable(int left, int right)
+{
+  return reg_update(MAX9867_15_MIC, 0x30, (!!left << 5) | (!!right << 4));
+}
+
+int max9867_digital_volume_slew(int slew_80)
+{
+  return reg_update(MAX9867_16_MODE, 0x80, !!slew_80 << 7);
+}
+
+int max9867_volume_change_smoothing(int one_step)
+{
+  return reg_update(MAX9867_16_MODE, 0x40, !!one_step << 6);
+}
+
+int max9867_linein_zero_crossing_disable(int disable)
+{
+  return reg_update(MAX9867_16_MODE, 0x20, !!disable << 5);
+}
+
+int max9867_jack_detect_enable(int enable)
+{
+  return reg_update(MAX9867_16_MODE, 0x08, !!enable << 3);
+}
+
+int max9867_headphone_mode(headphone_mode_t mode)
+{
+  return reg_update(MAX9867_16_MODE, 0x07, mode);
+}
+
+int max9867_power_enable(int shutdown, int enable)
+{
+  int err;
+
+  if (shutdown)
+    if ((err = reg_write(MAX9867_17_PWR_SYS, 0x00)))
```

Comment:
Is the intent if shutdown succeeds to move on to the 'enable' settings.  Seems like maybe the shutdown block should always return?